File: C++/题目1120：全排列.cpp

// 题目1120：全排列

/**
 * @author:wangzq
 * @email:wangzhenqing1008@163.com
 * @date:2015-06-30 11:01:53
 * @url:http://ac.jobdu.com/problem.php?pid=1120
 * 解题思路参考csdn:http://blog.csdn.net/u013027996/article/details/18735245
 */

#include <stdio.h>
#include <string>

using namespace std;
string array;
int len;

void swap(int a, int b) {
    char temp = array[b];
    array[b] = array[a];
    array[a] = temp;
}

void reverse(int a, int b) {
    while (a < b) {
        swap(a, b);
        a++;
        b--;
    }
}

void permutation() {

    int i, j;
    while (true) {
        printf("%s\n", array.c_str());
        for (i = len - 2; i >= 0; i--) {
            if (array[i] < array[i + 1]) {
                break;
            }
            if (i == 0) {
                return;
            }
        }
        for (j = len - 1; j > i; j--) {
            if (array[j] > array[i]) {
                break;
            }
        }

        swap(i, j);
        reverse(i + 1, len - 1);
    }
}


int main() {

    char data[10];
    while (scanf("%s", &data) != EOF) {
        array = data;
        len = array.size();
        if (len == 1) {
            printf("%s\n", array.c_str());
            printf("\n");
            continue;
        }
        permutation();
        printf("\n");
    }

    return 0;
}
/**************************************************************
Problem: 1120
User: wzqwsrf
Language: C++
Result: Accepted
Time:370 ms
Memory:1052 kb
****************************************************************/


                        

File: Java/题目1120：全排列.java

// 题目1120：全排列

/**
 * @author:wangzq
 * @email:wangzhenqing1008@163.com
 * @date:2015-06-30 11:01:53
 * @url:http://ac.jobdu.com/problem.php?pid=1120
 * 解题思路参考csdn:http://blog.csdn.net/u013027996/article/details/18735245
 */

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.StreamTokenizer;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

public class Main {

    private static List<String> arrList;

    public static void main(String[] args) throws Exception {
        StreamTokenizer st = new StreamTokenizer(
                new BufferedReader(new InputStreamReader(System.in)));
        while (st.nextToken() != StreamTokenizer.TT_EOF) {
            arrList = new ArrayList<String>();
            String a = st.sval;
            char array[] = a.toCharArray();
            Arrays.sort(array);
            permutation(array);
            Collections.sort(arrList);
            StringBuffer sb = new StringBuffer();
            for (int i = 0; i < arrList.size(); i++) {
                sb.append(arrList.get(i) + "\n");
            }
            System.out.println(sb);
        }
    }

    private static void permutation(char[] array) {
        // TODO Auto-generated method stub
        int len = array.length;
        while (true) {
            arrList.add(String.valueOf(array));
            int first = getFirst(array, len);
            if (first == -1) {
                return;
            }
            int i = len - 1;
            for (; i > first; i--) {
                if (array[i] > array[first]) {
                    break;
                }
            }
            swap(array, first, i);
            reverse(array, first + 1, len - 1);
        }
    }

    private static int getFirst(char[] array, int len) {
        for (int i = len - 2; i >= 0; i--) {
            if (array[i] < array[i + 1]) {
                return i;
            }
        }
        return -1;
    }

    private static void swap(char[] array, int a, int b) {
        char temp = array[b];
        array[b] = array[a];
        array[a] = temp;
    }

    private static void reverse(char[] array, int a, int b) {
        while (a < b) {
            swap(array, a, b);
            a++;
            b--;
        }
    }
}
/**************************************************************
 * Problem: 1120
 * User: wangzhenqing
 * Language: Java
 * Result: Accepted
 * Time:1700 ms
 * Memory:178508 kb
 ****************************************************************/


                        


